l3kernel/l3sys.dtx

```diff
@@ -361,6 +373,50 @@
 % \end{variable}
 % \end{macro}
 %
+% \begin{variable}{\c_sys_engine_exec_str,\c_sys_engine_format_str}
+%   Take the functions defined above, and set up the engine and format
+%   names.  \cs{c_sys_engine_exec_str} differs from \cs{c_sys_engine_str}
+%   as it is the \emph{actual} engine name, not a \enquote{filtered}
+%   version.  It differs for |ptex| and |uptex|, which have a leading
+%   |e|, and for |luatex|, because \LaTeX{} uses the \Lua HB\TeX{}
+%   engine.
+%
+%   \cs{c_sys_engine_format_str} is quite similar to
+%   \cs{c_sys_engine_str}, except that it differentiates |pdflatex| from
+%   |latex| (which is \pdfTeX{} in DVI mode).  This differentiation,
+%   however, is reliable only if the user doesn't change
+%   \cs{tex_pdfoutput:D} before loading this code.
+%    \begin{macrocode}
+\str_const:Nx \c_sys_engine_exec_str
+  {
+    \sys_if_engine_pdftex:T { pdf }
+    \sys_if_engine_xetex:T  { xe  }
+    \sys_if_engine_ptex:T   { ep  }
+    \sys_if_engine_uptex:T  { eup }
+    \sys_if_engine_luatex:T
+      {
+        lua \lua_now:e
+          {
+            if (pcall(require, 'luaharfbuzz')) then ~
+              tex.print("hb") ~
+            end
+          }
+      }
+    tex
+  }
+\str_const:Nx \c_sys_engine_format_str
```

Comment:
Doesn't this need some conditional code to deal with 'might not be LaTeX' cases, e.g. plain, ConTeXt, LuaMetaTeX, ...